router/dataplane_concurrency_model.gobra

```diff
@@ -0,0 +1,254 @@
+	acc(y.Place) && acc(y.State) && // existentials are expressed using ghost pointers
+	io.token(*y.Place) && dp.dp3s_iospec_ordered(*y.State, *y.Place) && 
+	ElemAuth((*y.State).ibuf, y.IBufY) && ElemAuth((*y.State).obuf, y.OBufY)
+}
+
+// initialize the shared invariant:
+ghost 
+requires io.token(p) && dp.dp3s_iospec_ordered(s, p)
+ensures  m.LockP() && m.LockInv() == SharedInv!< dp, y !>;
+func InitSharedInv(ghost dp io.DataPlaneSpec, ghost p io.Place, ghost s io.IO_dp3s_state_local) (ghost m *sync.Mutex, ghost y SharedArg) {
+	mV@ := sync.Mutex{}
+	m = &mV
+	pE@ := p
+	sE@ := s
+	yI := InitElemAuth(s.ibuf)
+	yO := InitElemAuth(s.obuf)
+	y := SharedArg{&pE, &sE, yI, yO}
+	fold SharedInv!< dp, y !>()
+	m.SetInv(SharedInv!< dp, y !>)
+}
+
+/////////////////////////////////////// Lib ///////////////////////////////////////
+
+pred Prophecy(int)
+
+ghost
+ensures 0 <= expectedM && Prophecy(expectedM)
+decreases
+func AllocProphecy() (expectedM int)
+
+/////////////////////////////////////// MultiBio ///////////////////////////////////////
+
+pred MultiReadBio(ghost t io.Place, n int) {
+  (n > 0 ==> io.CBioIO_bio3s_recv(t) && MultiReadBio(io.dp3s_iospec_bio3s_recv_T(t), n-1))
+}
+
+ghost
+requires m <= n && MultiReadBio(t, n)
+decreases m
+pure func MutliReadBio_next(ghost t io.Place, n, m int) (ghost tn io.Place) {
+  return m <= 0 ? t : unfolding MultiReadBio(t, n) in 
+    MutliReadBio_next(io.dp3s_iospec_bio3s_recv_T(t), n-1, m-1)
+}
+
+ghost 
+requires m <= n && MultiReadBio(t, n)
+decreases m
+pure func MutliReadBio_correctIfs(ghost t io.Place, n, m int, ghost k Key) bool {
+  return m <= 0 ? true : unfolding MultiReadBio(t, n) in 
+    match io.dp3s_iospec_bio3s_recv_R(t){
+      case io.IO_val_Unsupported{?ifs, _}:
+        k == ifs
+      case io.IO_val_Pkt2{?ifs, _ }:
+        k == ifs
+      default:
+        false
+    } && MutliReadBio_correctIfs(io.dp3s_iospec_bio3s_recv_T(t), n-1, m-1, k)
+}
+
+ghost
+requires io.CBioIO_bio3s_recv(t)
+decreases
+pure func IO_valToIO_Pkt3(ghost t io.Place) (ghost res seq[Elem]) {
+  return let pkt := io.dp3s_iospec_bio3s_recv_R(t) in
+    match pkt{
+      case io.IO_val_Pkt2{_, _}:
+        seq[Elem]{pkt.IO_val_Pkt2_2}
+      default:
+        seq[Elem]{}
+    }
+}
+
+ghost 
+requires  0 <= m && m <= n && MultiReadBio(t, n)
+ensures len(res) == m
+decreases m
+pure func MutliReadBio_IO_val(ghost t io.Place, n, m int) (ghost res seq[io.IO_val]) {
+  return m <= 0 ? seq[io.IO_val]{} : unfolding MultiReadBio(t, n) in 
+    seq[io.IO_val]{io.dp3s_iospec_bio3s_recv_R(t)} ++ MutliReadBio_IO_val(
+      io.dp3s_iospec_bio3s_recv_T(t), n-1, m-1)
+}
+
+ghost 
+requires m <= n && MultiReadBio(t,n)
+decreases m
+pure func MultiReadBio_Upd(ghost t io.Place, n, m int, ghost s io.IO_dp3s_state_local) io.IO_dp3s_state_local {
+  return m <= 0 ? s : unfolding MultiReadBio(t, n) in 
+    MultiReadBio_Upd(io.dp3s_iospec_bio3s_recv_T(t), n-1, m-1, add_ibuf(s, io.dp3s_iospec_bio3s_recv_R(t)))
+}
+
+ghost 
+decreases 
```

Comment:
I would have expected this function to have a precondition saying that `val` is either a `io.IO_val_Pkt2` or a `io.IO_val_Unsupported`

Reply:
Even though we only call this function with `io.IO_val_Pkt2` and `io.IO_val_Unsupported`, I would still use `match` instead of a precondition. I find it clearer because theoretically it could be called with other `io.IO_value`.

Reply:
Having a precondition here would allow us to find errors sooner if we try to pass sth other than `io.IO_val_Pkt2` or `io.IO_val_Unsupported` (having a precondition error is more noticeable than having confusing error messages later on because we are getting an `undefined` without noticing

Reply:
Should I remove the `default` case as well?

Reply:
you can keep it (I don't remember if Gobra complains if the match is not exhaustive)